File: 1/1-16.py
from typing import Union

def fast_expt_iter(b: Union[int, float], n: int, a:Union[int, float]) -> Union[int, float]:
    if n == 0:
        return a
    elif n % 2 == 0:
        return fast_expt_iter(b * b, n // 2, a)
    else:
        return fast_expt_iter(b, n - 1, a * b)

def fast_expt(b: Union[int, float], n: int) -> Union[int, float]:
    return fast_expt_iter(b, n, 1)

base = 2

exponent = 10

result = fast_expt(base, exponent)

print(f"{base}^{exponent} = {result}")

File: 1/1-16.hs
fastExptIter :: Double -> Int -> Double -> Double
fastExptIter b n a
    | n == 0    = a
    | even n    = fastExptIter (b * b) (n `div` 2) a
    | otherwise = fastExptIter b (n - 1) (a * b)

fastExpt :: Double -> Int -> Double
fastExpt b n = fastExptIter b n 1.0

main :: IO ()
main = do
    let base = 2.0
        exponent = 10
        result = fastExpt base exponent
    putStrLn $ show base ++ "^" ++ show exponent ++ " = " ++ show result